sqlx-core/src/pool/inner.rs

```diff
@@ -113,13 +113,15 @@ impl<DB: Database> SharedPool<DB> {
         let mut size = self.size();
 
         while size < self.options.max_connections {
-            let new_size = self.size.compare_and_swap(size, size + 1, Ordering::AcqRel);
-
-            if new_size == size {
-                return Some(DecrementSizeGuard::new(self));
+            if let Ok(new_size) =
+                self.size
+                    .compare_exchange(size, size + 1, Ordering::AcqRel, Ordering::Acquire)
+            {
+                if new_size == size {
+                    return Some(DecrementSizeGuard::new(self));
+                }
+                size = new_size;
```

Comment:
The semantics for `.compare_exchange()` state that the `Ok` value is always the same as the `current` parameter (in which case returning it is kinda redundant IMO but that's water under the bridge at this point).

This should be rewritten in one of two ways, take your pick (I think the `match` is more readable as I don't typically expect to see `if let Err(_)` with an `else`):

```rust
if let Err(new_size) = self.size.compare_exchange(size, size + 1, Ordering::AcqRel, Ordering::Acquire) {
    size = new_size;
} else {
    return Some(DecrementSizeGuard::new(self));
}
```

```rust
match self.size.compare_exchange(size, size + 1, Ordering::AcqRel, Ordering::Acquire) {
    Ok(_) => return Some(DecrementSizeGuard::new(self)),
    Err(new_size) => size = new_size,
}
```

Reply:
Thanks for catching an issue here.  Ok() does not contain the new value, but rather the previous value.
So the match expression is a fix in addition to being perhaps more customary.